fish-rust/src/signal.rs

```diff
@@ -65,3 +67,272 @@ pub fn signal_get_desc(sig: i32) -> &'static wstr {
+            Signal::SIGEMT => "Emulator trap",
+            Signal::SIGFPE => "Floating point exception",
+            Signal::SIGKILL => "Forced quit",
+            Signal::SIGBUS => "Misaligned address error",
+            Signal::SIGSEGV => "Address boundary error",
+            Signal::SIGSYS => "Bad system call",
+            Signal::SIGPIPE => "Broken pipe",
+            Signal::SIGALRM => "Timer expired",
+            Signal::SIGTERM => "Polite quit request",
+            Signal::SIGURG => "Urgent socket condition",
+            Signal::SIGSTOP => "Forced stop",
+            Signal::SIGTSTP => "Stop request from job control (^Z)",
+            Signal::SIGCONT => "Continue previously stopped process",
+            Signal::SIGCHLD => "Child process status changed",
+            Signal::SIGTTIN => "Stop from terminal input",
+            Signal::SIGTTOU => "Stop from terminal output",
+            Signal::SIGIO => "I/O on asynchronous file descriptior is possible",
+            Signal::SIGXCPU => "CPU time limit exceeded",
+            Signal::SIGXFSZ => "File size limit exceeded",
+            Signal::SIGVTALRM => "Virtual timer expired",
+            Signal::SIGPROF => "Profiling timer expired",
+            Signal::SIGWINCH => "Window size change",
+            #[cfg(any(target_os = "freebsd", target_os = "macos"))]
+            Signal::SIGINFO => "Information request",
+            Signal::SIGUSR1 => "User-defined signal 1",
+            Signal::SIGUSR2 => "User-defined signal 2",
+            #[cfg(any(target_os = "freebsd"))]
+            Signal::SIGTHR => "Thread interrupt",
+            #[cfg(any(target_os = "freebsd"))]
+            Signal::SIGLIBRT => "Real-time library interrupt",
+            #[cfg(target_os = "linux")]
+            Signal::SIGSTKFLT => "Stack fault",
+            #[cfg(target_os = "linux")]
+            Signal::SIGPWR => "Power failure",
+            Signal(_) => "Unknown",
+        }
+    }
+
+    pub const fn code(&self) -> i32 {
+        // Safety: Types share size and alignment but one just has additional semantic constraints
+        (unsafe { *(&self.0 as *const _ as *const u8) }) as i32
+    }
+
+    pub const fn new(sig: NonZeroU8) -> Self {
+        Signal(sig)
+    }
+
+    pub fn try_new(sig: i32) -> Option<Self> {
+        u8::try_from(sig)
+            .and_then(NonZeroU8::try_from)
+            .map(Signal::new)
+            .ok()
+    }
+}
+
+impl std::fmt::Debug for Signal {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.write_fmt(format_args!("{}({})", self.name(), self.code()))
+    }
+}
+
+impl std::fmt::Display for Signal {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        let name = self.name();
+        if name != Self::UNKNOWN_SIG_NAME {
+            f.write_str(name)
+        } else {
+            f.write_fmt(format_args!("Unrecognized Signal {}", self.code()))
+        }
+    }
+}
+
+impl From<NonZeroU8> for Signal {
```

Comment:
I think we can do without these int -> Signal `From` implementations. It's fine to make Signal explicit.